### ISSUE TITLE
Charset error when there are special characters in the file path

### ISSUE THREAD

**Issue by IlluminatiWave** [exiftool/exiftool#253]
Exiftool (12.82 / Win build) doesn't have the ability to read files if they have Chinese names,

```pwsh
cls; foreach ($charset in "Arabic","Baltic","Cyrillic","DOSCyrillic","DOSLatin1","DOSLatinUS","Greek","Hebrew","Latin","Latin2","MacCroatian","MacCyrillic","MacGreek","MacIceland","MacLatin2","MacRoman","MacRomanian","MacTurkish","Thai","Turkish","UTF8","Vietnam") { exiftool -charset $charset "Path\Chinese path篇\file.jpg" }
```

```cmd
Wildcards don't work in the directory specification
No matching files 
```

_Tested with chcp 936 (chinese) and with chcp 65001 (UTF8)_

**Comment by boardhead** [comment]
You should be setting "-charset filename=XXX" not "-charset XXX"

- Phil

**Comment by IlluminatiWave** [comment]
It didn't work, the program couldn't detect it.

```powershell
#This is only for changing the console printout
PS C:\Users\myUser\Downloads\test> function prompt {"PS $(Split-Path -leaf -path (Get-Location))> "}
PS test> cd .\合集\
#similar to "dir /b/s"
PS 合集> get-ChildItem | Select-Object -ExpandProperty Name
1.jpg
1.png
#I did some small tests with oxipng (check if the file is visible to other command line programs).
PS 合集> oxipng 1.png
Processing: 1.png
1.png: Could not optimize further, no change written
PS 合集> oxipng 1.jpg
Processing: 1.jpg
1.jpg: Invalid PNG header detected
# Try to read with exiftool
PS 合集>  exiftool .\1.jpg ;  exiftool ".\1.jpg" ;  exiftool 1.jpg ; exiftool "1.jpg"
Error: File not found - ./1.jpg
Error: File not found - ./1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
# loop "exiftool -charset "1.jpg"=XXX"
PS 合集> foreach ($charset in "Arabic","Baltic","Cyrillic","DOSCyrillic","DOSLatin1","DOSLatinUS","Greek","Hebrew","Latin","Latin2","MacCroatian","MacCyrillic","MacGreek","MacIceland","MacLatin2","MacRoman","MacRomanian","MacTurkish","Thai","Turkish","UTF8","Vietnam") { exiftool -charset "1.jpg"=$charset }
Invalid Charset 1.jpg
Error: File not found - Arabic
Invalid Charset 1.jpg
Error: File not found - Baltic
Invalid Charset 1.jpg
Error: File not found - Cyrillic
Invalid Charset 1.jpg
Error: File not found - DOSCyrillic
Invalid Charset 1.jpg
Error: File not found - DOSLatin1
Invalid Charset 1.jpg
Error: File not found - DOSLatinUS
Invalid Charset 1.jpg
Error: File not found - Greek
Invalid Charset 1.jpg
Error: File not found - Hebrew
Invalid Charset 1.jpg
Error: File not found - Latin
Invalid Charset 1.jpg
Error: File not found - Latin2
Invalid Charset 1.jpg
Error: File not found - MacCroatian
Invalid Charset 1.jpg
Error: File not found - MacCyrillic
Invalid Charset 1.jpg
Error: File not found - MacGreek
Invalid Charset 1.jpg
Error: File not found - MacIceland
Invalid Charset 1.jpg
Error: File not found - MacLatin2
Invalid Charset 1.jpg
Error: File not found - MacRoman
Invalid Charset 1.jpg
Error: File not found - MacRomanian
Invalid Charset 1.jpg
Error: File not found - MacTurkish
Invalid Charset 1.jpg
Error: File not found - Thai
Invalid Charset 1.jpg
Error: File not found - Turkish
Invalid Charset 1.jpg
Error: File not found - UTF8
Invalid Charset 1.jpg
Error: File not found - Vietnam
# Still doesn't detect Chinese charset
PS 合集> Test-Path -Path 1.jpg
True # Exist
PS 合集> Test-Path -Path 2.jpg
False # Doesn't exist
```

**Comment by boardhead** [comment]
"-charset filename=" is literal.

**Comment by boardhead** [comment]
Perhaps read FAQ 10: https://exiftool.org/faq.html#Q10

**Comment by StarGeekSpaceNerd** [comment]
Link for the [`-charset` option](https://exiftool.org/exiftool_pod.html#charset-TYPE-CHARSET).

If you look at your output, you see `Invalid Charset 1.jpg` which means exiftool is trying to `1.jpg` as the charset.  Then you get `Error: File not found - Arabic`, which means that exiftool is looking for a file called `Arabic`.

**Comment by MasterInQuestion** [comment]
Specifying &#x60;`-charset "FileName="`&#x60; should workaround which.
    Though the output shall be still tricky: if your shell wouldn't properly interpret UTF-8 (which is asserted).

**Comment by StarGeekSpaceNerd** [comment]
> The problem appears to be Windows only

Yes.  Mac/Linux have no problem dealing with UTF8 characters.  Also see [FAQ #18, problems with special characters on the Windows command line](https://exiftool.org/faq.html#Q18).  I ended up using the StackOverflow answer linked there. It causes problems with some older GUIs, but otherwise works.

> ExifTool asserted UTF-8 for filenames, while on Windows which appears to be UTF-16 (or else).

This is something that PowerShell is doing.  It thinks the output is ASCII and is "helping" you by converting it to UTF-16. See [Redirection/Pipe binary corruption with Windows PowerShell](https://exiftool.org/forum/index.php?topic=8137.msg41663#msg41663)

**Comment by MasterInQuestion** [comment]
UTF-16 filename encoding appears to be system-wide on Windows.
    PowerShell on Windows outputs (of console) default in UTF-16.
    [ As mentioned in: [https://stackoverflow.com/questions/57131654#comment100786572_57134096](https://stackoverflow.com/questions/57131654#comment100786572_57134096) ]

    Note that most Windows GUI options are actually stored in the Registry:
    Manipulating the Registry would be just equivalent as through GUI.

    And technically it breaks anything doesn't have proper Unicode support. (not only GUI programs)

**Comment by IlluminatiWave** [comment]
First suggestion (chcp 65001 - exiftool -charset filename=CHARSET)

```pwsh
PS 合集> chcp 65001
Página de códigos activa: 65001
PS 合集> Test-Path -Path 1.jpg; foreach ($charset in "Arabic","Baltic","Cyrillic","DOSCyrillic","DOSLatin1","DOSLatinUS","Greek","Hebrew","Latin","Latin2","MacCroatian","MacCyrillic","MacGreek","MacIceland","MacLatin2","MacRoman","MacRomanian","MacTurkish","Thai","Turkish","UTF8","Vietnam") { exiftool -charset "1.jpg"=$charset }
True
Invalid Charset 1.jpg
Error: File not found - Arabic
Invalid Charset 1.jpg
Error: File not found - Baltic
Invalid Charset 1.jpg
Error: File not found - Cyrillic
Invalid Charset 1.jpg
Error: File not found - DOSCyrillic
Invalid Charset 1.jpg
Error: File not found - DOSLatin1
Invalid Charset 1.jpg
Error: File not found - DOSLatinUS
Invalid Charset 1.jpg
Error: File not found - Greek
Invalid Charset 1.jpg
Error: File not found - Hebrew
Invalid Charset 1.jpg
Error: File not found - Latin
Invalid Charset 1.jpg
Error: File not found - Latin2
Invalid Charset 1.jpg
Error: File not found - MacCroatian
Invalid Charset 1.jpg
Error: File not found - MacCyrillic
Invalid Charset 1.jpg
Error: File not found - MacGreek
Invalid Charset 1.jpg
Error: File not found - MacIceland
Invalid Charset 1.jpg
Error: File not found - MacLatin2
Invalid Charset 1.jpg
Error: File not found - MacRoman
Invalid Charset 1.jpg
Error: File not found - MacRomanian
Invalid Charset 1.jpg
Error: File not found - MacTurkish
Invalid Charset 1.jpg
Error: File not found - Thai
Invalid Charset 1.jpg
Error: File not found - Turkish
Invalid Charset 1.jpg
Error: File not found - UTF8
Invalid Charset 1.jpg
Error: File not found - Vietnam
```

Second suggestion (chcp 65001 - exiftool -charset CHARSET=filename)

```pwsh
PS 合集> chcp 65001
Página de códigos activa: 65001
PS 合集> Test-Path -Path 1.jpg; foreach ($charset in "Arabic","Baltic","Cyrillic","DOSCyrillic","DOSLatin1","DOSLatinUS","Greek","Hebrew","Latin","Latin2","MacCroatian","MacCyrillic","MacGreek","MacIceland","MacLatin2","MacRoman","MacRomanian","MacTurkish","Thai","Turkish","UTF8","Vietnam") { exiftool -charset $charset="1.jpg" }
True
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
```

Third attempt (chcp 936, same testcode as second suggestion)

```pwsh
PS 合集> chcp 936
Página de códigos activa: 936 # Chinese charset console
PS 合集> Test-Path -Path 1.jpg; foreach ($charset in "Arabic","Baltic","Cyrillic","DOSCyrillic","DOSLatin1","DOSLatinUS","Greek","Hebrew","Latin","Latin2","MacCroatian","MacCyrillic","MacGreek","MacIceland","MacLatin2","MacRoman","MacRomanian","MacTurkish","Thai","Turkish","UTF8","Vietnam") { exiftool -charset $charset="1.jpg" }
True
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
Error: File not found - 1.jpg
```

In the end the correct code was to use exiftool -charset CHARSET=filename, but even so it is impossible for the program to recognize the file if any of the characters are Chinese (or other charset) in the folders.

```pwsh
PS альбом-компиляция> cd ..
PS test> chcp; Test-Path -Path C:\Users\myUser\Downloads\test\合集\1.jpg; foreach ($charset in "Arabic","Baltic","Cyrillic","DOSCyrillic","DOSLatin1","DOSLatinUS","Greek","Hebrew","Latin","Latin2","MacCroatian","MacCyrillic","MacGreek","MacIceland","MacLatin2","MacRoman","MacRomanian","MacTurkish","Thai","Turkish","UTF8","Vietnam") { exiftool -charset $charset="C:\Users\myUser\Downloads\test\合集\1.jpg" }
Página de códigos activa: 65001
True
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
PS test> chcp; Test-Path -Path C:\Users\myUser\Downloads\test\コンピレーション・アルバム\1.jpg; foreach ($charset in "Arabic","Baltic","Cyrillic","DOSCyrillic","DOSLatin1","DOSLatinUS","Greek","Hebrew","Latin","Latin2","MacCroatian","MacCyrillic","MacGreek","MacIceland","MacLatin2","MacRoman","MacRomanian","MacTurkish","Thai","Turkish","UTF8","Vietnam") { exiftool -charset $charset="C:\Users\myUser\Downloads\test\コンピレーション・アルバム\1.jpg" }
Página de códigos activa: 65001
True
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
PS test> chcp; Test-Path -Path C:\Users\myUser\Downloads\test\альбом-компиляция\1.jpg; foreach ($charset in "Arabic","Baltic","Cyrillic","DOSCyrillic","DOSLatin1","DOSLatinUS","Greek","Hebrew","Latin","Latin2","MacCroatian","MacCyrillic","MacGreek","MacIceland","MacLatin2","MacRoman","MacRomanian","MacTurkish","Thai","Turkish","UTF8","Vietnam") { exiftool -charset $charset="C:\Users\myUser\Downloads\test\альбом-компиляция\1.jpg" }
Página de códigos activa: 65001
True
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
Wildcards don't work in the directory specification
No matching files
```

Actually, I tried to do the opposite, to try to read the metadata from a folder with normal characters, but where the files are the ones with the special characters. Which although they are not represented correctly (because I didn't even apply a charset to them), exiftools is able to find them.

```pwsh
PS Orig> cd '.\compilation album\'
PS compilation album> exiftool -r .\
======== ./One.jpg
ExifTool Version Number         : 12.82
File Name                       : One.jpg
Directory                       : .
File Size                       : 3.7 MB
File Modification Date/Time     : 2024:02:06 03:37:37-06:00
File Access Date/Time           : 2024:04:19 15:32:57-06:00
File Creation Date/Time         : 2024:04:19 15:23:54-06:00
File Permissions                : -rw-rw-rw-
File Type                       : JPEG
File Type Extension             : jpg
MIME Type                       : image/jpeg
JFIF Version                    : 1.02
Resolution Unit                 : inches
X Resolution                    : 300
Y Resolution                    : 300
Comment                         : SYSTEMAX JFIF Encoder Ver.2.0
Image Width                     : 4188
Image Height                    : 4425
Encoding Process                : Baseline DCT, Huffman coding
Bits Per Sample                 : 8
Color Components                : 3
Y Cb Cr Sub Sampling            : YCbCr4:4:4 (1 1)
Image Size                      : 4188x4425
Megapixels                      : 18.5
======== ./ð×ð┤ð©ð¢.jpg # One in russian (Один.jpg)
ExifTool Version Number         : 12.82
File Name                       : ð×ð┤ð©ð¢.jpg
Directory                       : .
File Size                       : 3.7 MB
File Modification Date/Time     : 2024:02:06 03:37:37-06:00
File Access Date/Time           : 2024:04:19 15:32:57-06:00
File Creation Date/Time         : 2024:04:19 15:24:56-06:00
File Permissions                : -rw-rw-rw-
File Type                       : JPEG
File Type Extension             : jpg
MIME Type                       : image/jpeg
JFIF Version                    : 1.02
Resolution Unit                 : inches
X Resolution                    : 300
Y Resolution                    : 300
Comment                         : SYSTEMAX JFIF Encoder Ver.2.0
Image Width                     : 4188
Image Height                    : 4425
Encoding Process                : Baseline DCT, Huffman coding
Bits Per Sample                 : 8
Color Components                : 3
Y Cb Cr Sub Sampling            : YCbCr4:4:4 (1 1)
Image Size                      : 4188x4425
Megapixels                      : 18.5
======== ./õ©Ç.jpg # chinese/japanese one kanji (一.jpg)
ExifTool Version Number         : 12.82
File Name                       : õ©Ç.jpg
Directory                       : .
File Size                       : 3.7 MB
File Modification Date/Time     : 2024:02:06 03:37:37-06:00
File Access Date/Time           : 2024:04:19 15:32:57-06:00
File Creation Date/Time         : 2024:04:19 15:24:58-06:00
File Permissions                : -rw-rw-rw-
File Type                       : JPEG
File Type Extension             : jpg
MIME Type                       : image/jpeg
JFIF Version                    : 1.02
Resolution Unit                 : inches
X Resolution                    : 300
Y Resolution                    : 300
Comment                         : SYSTEMAX JFIF Encoder Ver.2.0
Image Width                     : 4188
Image Height                    : 4425
Encoding Process                : Baseline DCT, Huffman coding
Bits Per Sample                 : 8
Color Components                : 3
Y Cb Cr Sub Sampling            : YCbCr4:4:4 (1 1)
Image Size                      : 4188x4425
Megapixels                      : 18.5
    1 directories scanned
    3 image files read
```

That is, the error is encountered when the image path contains some special character between the parent folder names, not between the file name.

**Comment by StarGeekSpaceNerd** [comment]
Please take the time to read the [`-charset` option](https://exiftool.org/exiftool_pod.html#charset-TYPE-CHARSET) and the FAQ linked above. Every one of your tests uses an incorrect command.  The example Phil gave of `-charset filename=XXX` does not mean replacing `filename` with the filename.  The only thing replaced is the XXX with the character set you are using.

A correct command would be 

`exiftool -charset filename=Latin2 1.jpg`

**Comment by IlluminatiWave** [comment]
>-charset [[TYPE=]CHARSET]
If TYPE is or not specified, this option sets the ExifTool character encoding for output tag values when reading and input values when writing, with a default of . If no CHARSET is given, a list of available character sets is returned. Valid CHARSET values are:ExifToolUTF8

From what I understand, the charset command works to process the file content (i.e. the key : values), not to handle the parameter of image paths (folders with special characters), so the charset command isn't relevant for the error (all images have metadata in English/ it's the same image, but in different paths and different names, I don't have a problem with that)

```pwsh
PS compilation album> exiftool -r .\
======== ./ð×ð┤ð©ð¢.jpg
ExifTool Version Number         : 12.82
File Name                       : ð×ð┤ð©ð¢.jpg
Directory                       : .
File Size                       : 3.7 MB
File Modification Date/Time     : 2024:02:06 03:37:37-06:00
File Access Date/Time           : 2024:04:19 15:47:43-06:00
File Creation Date/Time         : 2024:04:19 15:24:56-06:00
File Permissions                : -rw-rw-rw-
File Type                       : JPEG
File Type Extension             : jpg
MIME Type                       : image/jpeg
JFIF Version                    : 1.02
Resolution Unit                 : inches
X Resolution                    : 300
Y Resolution                    : 300
Comment                         : SYSTEMAX JFIF Encoder Ver.2.0
Image Width                     : 4188
Image Height                    : 4425
Encoding Process                : Baseline DCT, Huffman coding
Bits Per Sample                 : 8
Color Components                : 3
Y Cb Cr Sub Sampling            : YCbCr4:4:4 (1 1)
Image Size                      : 4188x4425
Megapixels                      : 18.5
```
It is strange to say the least that it does not even work with names with special characters (using it recursively finds it, but using it directly fails to do so).

```pwsh
PS compilation album> chcp; Test-Path .\Один.jpg; exiftool -charset filename=DOSLatin1 .\Один.jpg
Página de códigos activa: 65001 
True
No matching files # same error with filename=UTF8
```

**Comment by MasterInQuestion** [comment]
[[
    for ( "Arabic", "Baltic", "Cyrillic", ... ) {
    exiftool -charset ${<span>@</span>1}="${Path}";
    };
]]
; alike really doesn't make sense: ${<span>@</span>1} is not valid TYPE (per the [doc](https://exiftool.org/exiftool_pod.html#charset-TYPE-CHARSET)).

    &#x60;`-charset "FileName="`&#x60; is to be specified literally.

**Comment by IlluminatiWave** [comment]
>     ``-charset "FileName="`` is to be specified literally.


true, the original script didn't make sense, but I've corrected it thanks to @StarGeekSpaceNerd's comment. 

after that and seeing some extra comments, I changed the command to something similar to 

 exiftool -charset $charset="1.jpg" 

where $charset is the character set allowed by exiftools (exiftool -charset UTF8="1.jpg" )

but this didn't help either

Now the next command I tried on a path without special characters (in the directories) was.

exiftool -charset filename=Cyrillic .\Один.jpg

but it still can't find it, however, trying to access it recursively seems to work

exiftool -r .\

even in cmd doesn't work

C:\Users and Downloads album> exiftool -charset filename=Cyrillic Один.jpg

**Comment by MasterInQuestion** [comment]
https://exiftool.org/exiftool_pod.html#WINDOWS-UNICODE-FILE-NAMES
    Use nothing (empty) as CHARSET. (pass the mentioned, literally)

**Comment by IlluminatiWave** [comment]
![image](https://github.com/exiftool/exiftool/assets/42191299/f84d4733-08ef-4fb4-8563-15faff1aa3f0)

one.jpg bruh

Leaving it empty didn't work either (I tried, but nobody seems to believe me :L).

**Comment by MasterInQuestion** [comment]
Try CMD?
    Also read my 2 [previous](https://github.com/exiftool/exiftool/issues/253#issuecomment-2063406000) comments about UTF-16.

    The workaround should only work to limited extent.
    &#x60;`-charset "FileName="`&#x60; hands off the filename handling to the OS: whatever happens later, out of the control.

**Comment by boardhead** [comment]
It seems like you are intentionally trying to waste my time:

> C:\Users and Downloads album> exiftool -charset filename=Cyrillic Один.jpg

This will only work if your shell encoding is Cyrillic, which is very unlikely.

> PS compilation album> chcp; Test-Path .\Один.jpg; exiftool -charset filename=UTF8 Оne.jpg

You have finally got the -charset option and the shell encoding correct, but for some reason you are running ExifTool on One.jpg instead of Один.jpg  ... Why?

- Phil

**Comment by IlluminatiWave** [comment]
really, I'm not trying to waste your time.

![image](https://github.com/exiftool/exiftool/assets/42191299/33e90bc9-98aa-4eba-90ad-871015a97ace)

chcp 1251 = Windows Cylliric
Test-Path -Path Один.jpg = ask if "Один.jpg" exist 

exiftool -charset filename=Cyrillic Один.jpg   =   The command should work, but it doesn't.

cmd...

![image](https://github.com/exiftool/exiftool/assets/42191299/daa5ef05-92b0-4058-aecf-9d8f00bca4c4)


trying this
>     ``-charset "FileName="`` is to be specified literally.

![image](https://github.com/exiftool/exiftool/assets/42191299/4cb8c9e2-8455-49dd-807c-21c00d995195)

The reason why I tried to use cilirico with an image that does not have any cilirico characters, is because I wanted to prove my point, exiftools for some reason cannot see files with special characters (cilirico, Chinese, Japanese).

neither using chcp 65001 (utf) nor using the special character encoding (jp = chcp 1250 I think, which for some reason exiftools uses latin for japanese instead of shift-jis / chcp 932, zh = 936, or ru = chcp 1251)... which now that I think about it, what happens with names that combine different charset types in the name

![image](https://github.com/exiftool/exiftool/assets/42191299/3e40e0a3-e0b1-4c90-b9e1-d7aa1e3723d5)

I have been for hours trying many things, none of them work, the only thing that has really worked for me, is to create a script that makes a copy in the C:/temp folder and process the image from there (because I have no special characters, and rename the image using hash, I just don't have special characters).

I didn't want to mention some things I had already tried because I thought it was pretty obvious they wouldn't work (-charset "FileName="), but in the end I had to try again to get the screenshots.

I've really tried everything users have suggested, either with chcp 65001, or using other character sets, but nothing has worked, command combinations, etc. I've gotten to the point where I'm doubting if it's my windows build failing or people really think I'm pulling their leg. 

Not even the help forum has helped me.

It all started with trying to read a file with Chinese text, but the error seems to propagate to any kind of name other than "latin/utf8".

Regarding the utf8 /16 names, as far as I can see (and it shows in the screenshots) both my pwsh console and my cmd console support Chinese/Russian characters, if this was not the case, then the console would show character aberrations (I usually see them quite a lot in Japanese zip, I must give them a treatment so it doesn't happen, this is not really relevant).

The last desperate solution I can think of is to change my whole system to Chinese/Russian/Japanese language and try my luck if exiftools is able to read this content.

PS: Just in case anyone thinks of it, no, changing the console font doesn't work either. 

![image](https://github.com/exiftool/exiftool/assets/42191299/a713cf7d-052a-45fb-a9d9-387bd67c49ac)

**Comment by MasterInQuestion** [comment]
Mere "[CHCP](https://www.google.com/search?hl=en&gl=ca&num=10&q=CMD+CHCP)" itself may not suffice: as which does not influence some further handling of Windows underhood.
    Try setting "HKLM\SYSTEM\CurrentControlSet\Control\Nls\CodePage": "OEMCP" and reboot. (much equivalent [ [https://stackoverflow.com/questions/57131654#57134096](https://stackoverflow.com/questions/57131654#57134096) ])
<img src="https://github.com/MasterInQuestion/attach/raw/main/exiftool/exiftool/253/中.webp" alt="    [ &quot;中.webp&quot; ]" />
    &#x60;exiftool&#x60; toolchain from: [ https://github.com/exiftool/exiftool/issues/260](https://github.com/exiftool/exiftool/issues/260)

    See also:
    https://stackoverflow.com/questions/56419639#68066008
    https://superuser.com/questions/472288#472338
    https://community.notepad-plus-plus.org/post/74626
    [ https://github.com/Disassembler0/Win10-Initial-Setup-Script/issues/252#issuecomment-576481981](https://github.com/Disassembler0/Win10-Initial-Setup-Script/issues/252#issuecomment-576481981)
    https://www.google.com/search?hl=en&gl=ca&num=100&q=%22Nls%22+%22CodePage%22+%22ACP%22|%22OEMCP%22

**Comment by IlluminatiWave** [comment]
OH thanks. it worked or activating the Beta charset.

I had already tried changing the language to the target character system (Chinese > windows Chinese, Japanese > Japanese. etc.) although it worked relatively well, it was still incompatible Chinese/Japanese with Cyrillic characters (a bit annoying really, I got a bit lost when I changed the languages xd).

While it is true that I had already taken into account the utf16 issue (try to open cmd with different character systems chcp ~ /k or enable compatibility mode in cmd). what really solved the problem was to enable beta compatibility. I had ruled out at the time that it was a console thing because some tools like Oxipng, Ffmpeg or even Python itself (as I had made a script to serve as an intermediary with Exiftools when it found paths with special characters o.O) was the problem (moreover, I even already knew about this error, as with shift-jis encoded Zips I have to "prepare" them to extract them without the characters coming out corrupted).

Using character compatibility even makes it possible to read the metadata without even passing commands for extra encoding ([test image](https://hanasakahaha.com/wp-content/uploads/2018/03/%E8%99%B9.jpg) / exiftool 虹.jpg (chcp 65001).


![image](https://github.com/exiftool/exiftool/assets/42191299/fd5ab847-90fb-4177-9ac5-8e8dd822cd34)


The "problem" could be considered as solved, although I don't know, I have my doubts whether to consider this a windows problem or an exiftools compatibility problem with character path decoding. Anyway, it's good to know that there are alternatives to chcp, change the system language or even locale emulator (yes, I tried this too xd).
